### PR TITLE
Fixed #28540 -- Checked readability of file uploaded through TemporaryUploadedFile

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

**Comment by timgraham** [comment]
Could you add some more code comments explaining the implementation? I have trouble reviewing as I'm not sure what the intention of each line is. The test seems to belong in `tests/file_storage` since `FileFIeld` isn't modified by this patch.

**Comment by claudep** [comment]
Carlton, I don't share your opinion, but it may be that my code is not doing what I meant. My intent is to try guessing the OS default permissions if we created a new file in the target folder in the first place, then applying those permissions to the moved file. Creating a new file in that folder or moving a file from a temporary location to that folder is an implementation detail and shouldn't affect the final file permissions.

**Comment by carltongibson** [comment]
> Creating a new file in that folder or moving a file from a temporary location to that folder is an implementation detail and shouldn't affect the final file permissions.

That I do agree with. The weirdness here is that the behaviour isn't consistent (i.e. that `FileField` isn't _always_ using a tmp file — I haven't looked at that yet.) 

That aside, my concern here is that we're (implicitly?) doing something that simply isn't our business. (Maybe I'm wrong about that.) Better, for me, would be a check or a warning or a error pointing to `FILE_UPLOAD_PERMISSIONS`, so at least it's opt-in on the user's part. (Not 100% against: just my feeling looking at it.)

> ...it may be that my code is not doing what I meant...

I think it is. 

```
>>> import os
>>> cur_umask = os.umask(0)
>>> os.umask(cur_umask)
0
>>> file_perms = 0o666 & ~cur_umask
>>> oct(file_perms)
'0o644' 
```

**Comment by claudep** [comment]
I see your point. Read also the original optimization ticket: https://code.djangoproject.com/ticket/27334. Before that, the behavior was more consistent but performed poorer.

**Comment by claudep** [comment]
I tried to address Tim's comments. Carlton, how do you see this going forward? Would you suggest exploring a different fix?

**Comment by carltongibson** [comment]
Hi @claudep. Thanks for the work. 

> Would you suggest exploring a different fix?

The one I have in mind, but haven't yet had the chance to look into, is whether it makes sense for `FileField` to always use the _move_ strategy, rather than _copy_ for small files and _move_ for large files. 
(This is the underlying behaviour that the small change in https://code.djangoproject.com/ticket/27334 leveraged.) 

I don't currently know if that would be at all feasible. (I need to look. 🙂)

Independently of that, I'm not sure (either way) whether we should address this directly (vs just documenting it). Guessing the permissions to use (still) feels to me like something we should refuse to do. (Happy though if the consensus is the other way.)

There doesn't seem to be any easy workaround; there's no option to `tempfile....`: 

https://github.com/python/cpython/blob/12a08c47601cadea8e7d3808502cdbcca87b2ce2/Lib/tempfile.py#L258

(I don't know that the doc's "On most platforms, temporary files will have a mode of 0o600..." is quite as helpful as it could be.)

tl;dr: all a bit non-commital at this moment.

**Comment by claudep** [comment]
The 0o600 perms for temp file are a security measure taken because the tmp directory is shared to all users on the system (a good thing). IMHO we don't have this concern for final upload locations. So having all uploaded files with the 0o600 tmp perms is not an option for me, even if documented.

**Comment by carltongibson** [comment]
Right. So how about a system check that `FILE_UPLOAD_PERMISSIONS` is set, which I can silence if I really do want `None`, and which will remind me otherwise?

I always want `0o640` rather than `0o644`. No doubt others have their own preferences. My concern is that in guessing we get involved in a complex area that really isn't our concern. I'd much rather force the user to consider it.

**Comment by carltongibson** [comment]
Hey @claudep — What's your thought here then? Do you think a check pointing users to `FILE_UPLOAD_PERMISSIONS` would be sufficient, or do you (still) prefer that we try to handle it directly? (We've discussed it and I'm happy to go either way after that. If it's the latter I'll look again at the actual implementation.) Thanks!

**Comment by timgraham** [comment]
The current behavior (if a bit weird) seems consistent with [the documentation](https://docs.djangoproject.com/en/dev/ref/settings/), "On most platforms, temporary files will have a mode of 0o600, and files saved from memory will be saved using the system’s standard umask." By default, `MemoryFileUploadHandler` is used for files up to `settings.FILE_UPLOAD_MAX_MEMORY_SIZE`, otherwise `TemporaryFileUploadHandler` is used.

If we decide not to make a change (probably the discussion should move to django-developers), then we could at least add a note to the [deployment checklist](https://docs.djangoproject.com/en/dev/howto/deployment/checklist/). A system check feels a bit heavy handed as none of the open source Django projects I checked have specified this setting.

**Comment by claudep** [comment]
ML thread: https://groups.google.com/forum/#!topic/django-developers/h9XbQAPv5-I

**Comment by carltongibson** [comment]
Thanks @claudep! 👍

**Comment by carltongibson** [comment]
OK, I've opened #10194 with Tim's suggestion here. 

Whilst looking into I've become more convinced this is too complex a matter for us to just silently impose a behaviour. 

If I've used the more esoteric [File Upload Settings](https://docs.djangoproject.com/en/2.0/ref/settings/#file-uploads) to customise uploads (say, a custom handler writing temporary files to a (safe) process specific tmp folder, with known permissions, not matching the process `umask`...and so on...) then I **can't** have Django just resetting the permissions after the upload handler has done it's thing. 

So I start to think of all number of weird and wonderful ways it could go wrong...

Whereas it all becomes moot if the user just sets `FILE_UPLOAD_PERMISSIONS`. 

Hence I'd be :-1: on this proposal at this point.

**Comment by claudep** [comment]
As for me, I consider having two different file permissions for the default upload mechanism is a bug. We can document it, document workarounds, but that does not change the fact.

**Comment by claudep** [comment]
> Give `FILE_UPLOAD_PERMISSIONS` a better default value than `None`.

I thought about setting it to 0o644 or 0o640, but this has the drawback to not honor the  current umask.

> Reduce default `FILE_UPLOAD_HANDLERS` to a single class.

Not an option IMHO.

> Make TemporaryFileUploadHandler set a better permission after the move.

I thought that's what my patch was doing, isn't it :-)

**Comment by carltongibson** [comment]
> ... has the drawback to not honor the current umask.

Chances are I'm not using the umask to control the permissions for uploaded files unless I've thought about it. (_Do I want then the same as for other files my process creates?_) If I've thought about it, then I can set it back to `None`. I think that might be OK (better even) as a default. (Discuss 🙂)

> Not an option IMHO.

Fine. 

> I thought that's what my patch was doing

Well yes, but:

* Here, I was thinking as configurable (i.e. overridable) as an attribute on the upload handler rather than in the storage. (We'd need some opt-out...) 
* I was hoping to avoid the umask logic entirely: I think we're misrepresenting our user if we think they'll have correctly set the process umask but won't have/can't just provide a value to `FILE_UPLOAD_PERMISSIONS`.

As such, I kind of see option three as just an over-elaborate, and probably inferior version of option one ("Give `FILE_UPLOAD_PERMISSIONS` a better default value".)

File: django/core/files/storage.py [hunk 1]
```diff
@@ -285,6 +285,15 @@ def _save(self, name, content):
 
         if self.file_permissions_mode is not None:
             os.chmod(full_path, self.file_permissions_mode)
+        else:
+            # If the file was moved, check that permissions are like those of a
+            # newly created file. We determine first the 'should-be' permissions
+            # for new files, then update the file perms if they don't match.
+            cur_umask = os.umask(0)
+            os.umask(cur_umask)
```
Comment (adamchainz) [inline django/core/files/storage.py]: Of separate concern to the discussion so far, but I'd like to point out the [linux umask manpage](http://man7.org/linux/man-pages/man2/umask.2.html) notes:

> It is impossible to use umask() to fetch a process's umask without at
> the same time changing it.  A second call to umask() would then be
> needed to restore the umask.  The nonatomicity of these two steps
> provides the potential for races in multithreaded programs.

> Since Linux 4.7, the umask of any process can be viewed via the Umask
> field of /proc/[pid]/status.  Inspecting this field in
> /proc/self/status allows a process to retrieve its umask without at
> the same time changing it.

I don't like the sound of races in multithreaded programs, a lot of users run multithreaded django servers. Also presumably that means lines 233-237 are already vulnerable to this race condition.

**Comment by timgraham** [comment]
The default value of `FILE_UPLOAD_PERMISSION` was changed to 0o644 in  22aab8662f0368b63f91f2526bdd0532524bc0fe.
